src/darwin/Framework/CHIPTests/MTROTAProviderTests.m

```diff
@@ -888,7 +925,482 @@ - (void)test005_DoBDXTransferAllowUpdateRequest
+            NSData * data = [readHandle readDataUpToLength:blockSize.unsignedLongValue error:&readError];
+            XCTAssertNil(readError);
+            XCTAssertNotNil(data);
+
+            BOOL isEOF = offset + blockSize.unsignedLongValue >= imageSize;
+
+            ++lastBlockIndex;
+
+            if (isEOF) {
+                sOTAProviderDelegate.blockQueryHandler = nil;
+            }
+
+            completion(data, isEOF);
+
+            if (isEOF) {
+                [bdxQueryExpectation2 fulfill];
+            }
+        };
+        sOTAProviderDelegate.transferEndHandler
+            = ^(NSNumber * nodeID, MTRDeviceController * controller, NSError * _Nullable error) {
+                  XCTAssertEqualObjects(nodeID, @(kDeviceId2));
+                  XCTAssertEqual(controller, sController);
+                  XCTAssertNil(error);
+                  sOTAProviderDelegate.transferEndHandler = nil;
+                  [bdxEndExpectation2 fulfill];
+              };
+    };
+
+    // Flags to track if device1 and device have notified that the update has been applied.
+    static bool hasDevice1NotifiedUpdateApplied = false;
+    static bool hasDevice2NotifiedUpdateApplied = false;
+
+    // Set up the apply update request and update applied handlers for device1 and device2. Use the nodeID to verify which
+    // device send the applyUpdateRequest and subsequent notifyUpdateApplied methods.
+    sOTAProviderDelegate.applyUpdateRequestHandler = ^(NSNumber * nodeID, MTRDeviceController * controller,
+        MTROTASoftwareUpdateProviderClusterApplyUpdateRequestParams * params, ApplyUpdateRequestCompletion completion) {
+        XCTAssertTrue([nodeID isEqualToNumber:@(kDeviceId1)] || [nodeID isEqualToNumber:@(kDeviceId2)]);
+
+        bool isDeviceID1 = [nodeID isEqualToNumber:@(kDeviceId1)];
+        NSData * updateToken = (isDeviceID1) ? updateToken1 : updateToken2;
+
+        // Device1 is updated to version 10 and device2 to version 5.
+        NSNumber * kSoftwareVersion = (isDeviceID1) ? kUpdatedSoftwareVersion_10 : kUpdatedSoftwareVersion_5;
+        NSString * otaImageFilePath = (isDeviceID1) ? otaRawImagePath1 : otaRawImagePath2;
+        NSString * otaDownloadedFilePath = (isDeviceID1) ? kOtaDownloadedFilePath1 : kOtaDownloadedFilePath2;
+
+        XCTAssertEqual(controller, sController);
+        XCTAssertEqualObjects(params.updateToken, updateToken);
+        XCTAssertEqualObjects(params.newVersion, kSoftwareVersion);
+
+        XCTAssertTrue([[NSFileManager defaultManager] contentsEqualAtPath:otaImageFilePath andPath:otaDownloadedFilePath]);
+
+        [sOTAProviderDelegate respondToApplyUpdateRequestWithAction:MTROTASoftwareUpdateProviderOTAApplyUpdateActionProceed
+                                                         completion:completion];
+
+        if (isDeviceID1) {
+            [applyUpdateRequestExpectation1 fulfill];
+        } else {
+            [applyUpdateRequestExpectation2 fulfill];
+        }
+    };
```

Comment:
We need to `sOTAProviderDelegate.applyUpdateRequestHandler = nil;` once both ApplyUpdateRequest commands have been received, no?  So need to keep a `__block` counter of how many calls we got, or booleans for separate devices or something.

Reply:
i thought to reset both applyUpdateRequestHandler and notifyUpdateAppliedHandler together to save on the booleans which is silly. will do it. makes sense

Reply:
done.